saleor/graphql/order/mutations/orders.py

```diff
@@ -574,3 +607,219 @@ def perform_mutation(cls, root, info, **data):
+                        code=OrderErrorCode.ZERO_QUANTITY,
+                        params={"variants": invalid_ids},
+                    ),
+                }
+            )
+        return lines_to_add
+
+    @staticmethod
+    def validate_variants(order, variants):
+        try:
+            channel = order.channel
+            validate_product_is_published_in_channel(variants, channel)
+            validate_variant_channel_listings(variants, channel)
+        except ValidationError as error:
+            raise ValidationError({"input": error})
+
+    @staticmethod
+    def add_lines_to_order(order, lines_to_add, manager):
+        try:
+            return [
+                add_variant_to_order(
+                    order,
+                    variant,
+                    quantity,
+                    manager,
+                    allocate_stock=order.is_unconfirmed(),
+                )
+                for quantity, variant in lines_to_add
+            ]
+        except TaxError as tax_error:
+            raise ValidationError(
+                "Unable to calculate taxes - %s" % str(tax_error),
+                code=OrderErrorCode.TAX_ERROR.value,
+            )
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        order = cls.get_node_or_error(info, data.get("id"), only_type=Order)
+        cls.validate_order(order)
+        lines_to_add = cls.validate_lines(info, data)
+        variants = [line[1] for line in lines_to_add]
+        cls.validate_variants(order, variants)
+
+        lines = cls.add_lines_to_order(order, lines_to_add, info.context.plugins)
+
+        # Create the products added event
+        events.order_added_products_event(
+            order=order, user=info.context.user, order_lines=lines_to_add
+        )
+
+        recalculate_order(order)
+        return OrderLinesCreate(order=order, order_lines=lines)
+
+
+class OrderLineDelete(EditableOrderValidationMixin, BaseMutation):
+    order = graphene.Field(Order, description="A related order.")
+    order_line = graphene.Field(
+        OrderLine, description="An order line that was deleted."
+    )
+
+    class Arguments:
+        id = graphene.ID(description="ID of the order line to delete.", required=True)
+
+    class Meta:
+        description = "Deletes an order line from an order."
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+        error_type_class = OrderError
+        error_type_field = "order_errors"
+
+    @classmethod
+    def perform_mutation(cls, _root, info, id):
+        line = cls.get_node_or_error(
+            info,
+            id,
+            only_type=OrderLine,
+        )
+        order = line.order
+        cls.validate_order(line.order)
+
+        db_id = line.id
+        line_info = OrderLineData(
+            line=line,
+            quantity=line.quantity,
+            variant=line.variant,
+            warehouse_pk=line.allocations.first().stock.warehouse.pk,
```

Comment:
If you perform this mutation for draft order then the line doesn't have any allocation. We should fetch `warehouse_pk` only for unconfirmed orders.